File: java-spring/architecture-shell-cqrs/src/main/java/com/architecture/shell/cqrs/behaviors/TelemetryBehavior.java
package com.architecture.shell.cqrs.behaviors;

import com.architecture.shell.cqrs.BaseRequest;
import com.architecture.shell.cqrs.PipelineBehavior;
import com.architecture.shell.cqrs.RequestHandlerDelegate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Pipeline behavior for telemetry and logging.
 * Logs request type, duration, status, and exceptions.
 *
 * Recommended Order: 40 (after transaction, measures handler execution)
 */
public class TelemetryBehavior<TRequest extends BaseRequest, TResponse>
    implements PipelineBehavior<TRequest, TResponse> {

    private static final Logger logger = LoggerFactory.getLogger(TelemetryBehavior.class);

    @Override
    public TResponse handle(TRequest request, RequestHandlerDelegate<TResponse> next) {
        long startTime = System.currentTimeMillis();
        logger.info("Request started: {}", request.getClass().getSimpleName());

        try {
            TResponse response = next.handle();
            long duration = System.currentTimeMillis() - startTime;
            logger.info("Request completed: {} Duration={}ms Status=Success",
                request.getClass().getSimpleName(), duration);
            return response;
        } catch (Exception ex) {
            long duration = System.currentTimeMillis() - startTime;
            logger.error("Request failed: {} Duration={}ms Status=Error Exception={}",
                request.getClass().getSimpleName(), duration, ex.getMessage(), ex);
            throw ex;
        }
    }
}

File: csharp-dotnet/src/Architecture.Shell.Cqrs/Behaviors/TelemetryBehavior.cs
// Copyright (c) 2025 Universal DDD Architecture. All rights reserved.

using System;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;

namespace Architecture.Shell.Cqrs.Behaviors;

/// <summary>
/// Pipeline behavior that logs telemetry for request processing.
/// Executes at order 40 (after transaction, wrapping handler execution) to measure handler duration.
/// </summary>
/// <typeparam name="TRequest">The request type being processed.</typeparam>
/// <typeparam name="TResponse">The response type from the handler.</typeparam>
/// <remarks>
/// Logs request type, duration, status, and exceptions per NFR-002 requirements.
/// For commands, includes TransactionId for correlation (BR-002).
/// </remarks>
public sealed class TelemetryBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
    where TRequest : IBaseRequest
{
    private readonly ILogger<TelemetryBehavior<TRequest, TResponse>> _logger;

    /// <summary>
    /// Gets the execution order for this behavior (40 - after transaction, wrapping handler).
    /// </summary>
    public int Order => 40;

    /// <summary>
    /// Initializes a new instance of the <see cref="TelemetryBehavior{TRequest, TResponse}"/> class.
    /// </summary>
    /// <param name="logger">Logger for telemetry output.</param>
    public TelemetryBehavior(ILogger<TelemetryBehavior<TRequest, TResponse>> logger)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    /// <summary>
    /// Logs telemetry before and after request processing.
    /// </summary>
    /// <param name="request">The request being processed.</param>
    /// <param name="continuation">The next behavior or handler in the pipeline.</param>
    /// <param name="cancellationToken">Cancellation token to cancel the operation.</param>
    /// <returns>The response from the handler.</returns>
    public async Task<TResponse> HandleAsync(
        TRequest request,
        RequestHandlerDelegate<TResponse> continuation,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);
        ArgumentNullException.ThrowIfNull(continuation);

        var requestType = typeof(TRequest).FullName ?? typeof(TRequest).Name;
        var stopwatch = Stopwatch.StartNew();

        try
        {
            // Given: Request starts processing
            _logger.LogInformation(
                "{RequestType} started at {Timestamp}",
                requestType,
                DateTimeOffset.UtcNow);

            // When: Execute handler
            var response = await continuation().ConfigureAwait(false);

            // Then: Log success telemetry
            stopwatch.Stop();
            _logger.LogInformation(
                "{RequestType} completed successfully - Duration: {Duration}ms, Status: Success",
                requestType,
                stopwatch.ElapsedMilliseconds);

            return response;
        }
        catch (OperationCanceledException)
        {
            // Cancellation requested
            stopwatch.Stop();
            _logger.LogWarning(
                "{RequestType} cancelled - Duration: {Duration}ms, Status: Cancelled",
                requestType,
                stopwatch.ElapsedMilliseconds);
            throw;
        }
        catch (Exception ex)
        {
            // Infrastructure or business error
            stopwatch.Stop();
            _logger.LogError(
                ex,
                "{RequestType} failed - Duration: {Duration}ms, Status: Error, Exception: {ExceptionType}",
                requestType,
                stopwatch.ElapsedMilliseconds,
                ex.GetType().Name);
            throw;
        }
    }
}
